packages/components/src/number-control/index.tsx

```diff
@@ -60,42 +59,24 @@ function UnforwardedNumberControl(
 		spinControls = 'none';
 	}
 
+	if ( typeof valueProp === 'number' ) {
+		// TODO: deprecate `value` as a `number`
+	}
+
+	const valuePropAsString =
+		valueProp !== undefined ? ensureString( valueProp ) : undefined;
+	const shiftStepAsNumber = ensureNumber( shiftStep );
+
 	const inputRef = useRef< HTMLInputElement >();
 	const mergedRef = useMergeRefs( [ inputRef, forwardedRef ] );
 
 	const isStepAny = step === 'any';
 	const baseStep = isStepAny ? 1 : ensureNumber( step );
 	const baseValue = roundClamp( 0, min, max, baseStep );
-	const constrainValue = (
-		value: number | string,
-		stepOverride?: number
-	) => {
-		// When step is "any" clamp the value, otherwise round and clamp it.
-		return isStepAny
-			? Math.min( max, Math.max( min, ensureNumber( value ) ) )
-			: roundClamp( value, min, max, stepOverride ?? baseStep );
-	};
 
 	const autoComplete = typeProp === 'number' ? 'off' : undefined;
 	const classes = classNames( 'components-number-control', className );
 
-	const spinValue = (
-		value: string | number | undefined,
-		direction: 'up' | 'down',
-		event: KeyboardEvent | MouseEvent | undefined
-	) => {
-		event?.preventDefault();
-		const shift = event?.shiftKey && isShiftStepEnabled;
-		const delta = shift ? ensureNumber( shiftStep ) * baseStep : baseStep;
-		let nextValue = isValueEmpty( value ) ? baseValue : value;
-		if ( direction === 'up' ) {
-			nextValue = add( nextValue, delta );
-		} else if ( direction === 'down' ) {
-			nextValue = subtract( nextValue, delta );
-		}
-		return constrainValue( nextValue, shift ? delta : undefined );
-	};
-
```

Comment:
Moved `constrainValue` and `spinValue` to a separate utils file — it makes more verbose, but they should be now pure functions , which makes them less tangled and easier to iterate on if necessary

Reply:
This doesn't hold true anymore (see [this convo](https://github.com/WordPress/gutenberg/pull/45982#discussion_r1039134411))